File: android/app/src/main/java/com/openburnbar/data/policy/MobileEscrowEnvelopePolicy.kt
package com.openburnbar.data.policy

enum class MobileEscrowImportFailure(val wire: String, val userVisibleLabel: String) {
    WRONG_DEVICE("wrong-device", "This envelope is for a different device"),
    EXPIRED_GRANT("expired-grant", "The transfer grant has expired"),
    REVOKED_GRANT("revoked-grant", "The transfer grant was revoked"),
    MISSING_KEY("missing-key", "This device is missing the escrow key"),
    MALFORMED_ENVELOPE("malformed-envelope", "The envelope is malformed"),
}

object MobileEscrowEnvelopePolicy {
    fun classify(
        targetDeviceId: String?,
        currentDeviceId: String?,
        grantStatus: String?,
        grantExpiresAtMs: Long?,
        nowMs: Long,
        hasPrivateKey: Boolean,
        envelopeWellFormed: Boolean,
    ): MobileEscrowImportFailure? {
        if (!envelopeWellFormed) return MobileEscrowImportFailure.MALFORMED_ENVELOPE
        if (!hasPrivateKey) return MobileEscrowImportFailure.MISSING_KEY
        val target = targetDeviceId?.trim().orEmpty()
        val current = currentDeviceId?.trim().orEmpty()
        if (target.isEmpty() || current.isEmpty() || target != current) {
            return MobileEscrowImportFailure.WRONG_DEVICE
        }
        if ((grantStatus ?: "").lowercase() == "revoked") return MobileEscrowImportFailure.REVOKED_GRANT
        val expiry = grantExpiresAtMs
        if (expiry == null || expiry <= 0L || expiry <= nowMs) {
            return MobileEscrowImportFailure.EXPIRED_GRANT
        }
        return null
    }
}

File: OpenBurnBarCore/Sources/OpenBurnBarMobilePolicy/MobileEscrowEnvelopePolicy.swift
import Foundation

/// Escrow import failures. Physical-device transfer stays a blocked ledger row
/// — this classifier is unit/scrutiny only.
public enum MobileEscrowImportFailure: String, Sendable, Equatable, CaseIterable {
    case wrongDevice = "wrong-device"
    case expiredGrant = "expired-grant"
    case revokedGrant = "revoked-grant"
    case missingKey = "missing-key"
    case malformedEnvelope = "malformed-envelope"

    public var userVisibleLabel: String {
        switch self {
        case .wrongDevice: return "This envelope is for a different device"
        case .expiredGrant: return "The transfer grant has expired"
        case .revokedGrant: return "The transfer grant was revoked"
        case .missingKey: return "This device is missing the escrow key"
        case .malformedEnvelope: return "The envelope is malformed"
        }
    }
}

public enum MobileEscrowEnvelopePolicy {
    public static func classify(
        targetDeviceId: String?,
        currentDeviceId: String?,
        grantStatus: String?,
        grantExpiresAtMs: Int64?,
        nowMs: Int64,
        hasPrivateKey: Bool,
        envelopeWellFormed: Bool
    ) -> MobileEscrowImportFailure? {
        guard envelopeWellFormed else { return .malformedEnvelope }
        guard hasPrivateKey else { return .missingKey }
        let target = targetDeviceId?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let current = currentDeviceId?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        guard !target.isEmpty, !current.isEmpty, target == current else {
            return .wrongDevice
        }
        let status = (grantStatus ?? "").lowercased()
        if status == "revoked" { return .revokedGrant }
        // Grants are allowed to omit expiry — firestore.rules does not persist
        // expiresAtMillis today. Only a present, elapsed timestamp expires.
        if let expiry = grantExpiresAtMs, expiry > 0, expiry <= nowMs {
            return .expiredGrant
        }
        return nil
    }
}
